File: ecs-hosted/go/application/network/pulumiInternalLoadBalancer.go
package network

import (
	"fmt"

	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/lb"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

/*
Create a internal (non-internet routable) NLB that will reside in a private subnet and be used to provide static IP address endpoint for our API
*/
func NewPulumiInternalLoadBalancer(ctx *pulumi.Context, name string, args *LoadBalancerArgs, opts ...pulumi.ResourceOption) (*PulumiInternalLoadBalancer, error) {
	var resource PulumiInternalLoadBalancer

	err := ctx.RegisterComponentResource("pulumi:internalLoadBalancer", name, &resource, opts...)
	if err != nil {
		return nil, err
	}

	// create our parented options
	options := append(opts, pulumi.Parent(&resource), pulumi.DeleteBeforeReplace(true))

	lbName := fmt.Sprintf("%s-lb", name)
	resource.LoadBalancer, err = lb.NewLoadBalancer(ctx, lbName, &lb.LoadBalancerArgs{
		LoadBalancerType: pulumi.String("network"),
		Internal:         pulumi.Bool(true),
		Subnets:          args.PrivateSubnetIds,
		IdleTimeout:      pulumi.Int(args.IdleTimeout),
		IpAddressType:    pulumi.String("ipv4"),
	}, options...)

	if err != nil {
		return nil, err
	}

	return &resource, nil
}

func (l *PulumiInternalLoadBalancer) CreateListener(ctx *pulumi.Context, name string, tgArn pulumi.StringOutput, certArn string, opts ...pulumi.ResourceOption) (*lb.Listener, error) {

	args := &lb.ListenerArgs{
		LoadBalancerArn: l.LoadBalancer.Arn,
		Port:            pulumi.Int(80),
		Protocol:        pulumi.String("TCP"),
		DefaultActions: &lb.ListenerDefaultActionArray{
			lb.ListenerDefaultActionArgs{
				TargetGroupArn: tgArn,
				Type:           pulumi.String("forward"),
			},
		},
	}

	if certArn != "" {
		args.CertificateArn = pulumi.String(certArn)
		args.SslPolicy = pulumi.String("ELBSecurityPolicy-TLS-1-2-2017-01")
		args.Protocol = pulumi.String("TLS")
		args.Port = pulumi.Int(443)
	}

	listOpts := append(opts, pulumi.DeleteBeforeReplace(true))
	listener, err := lb.NewListener(ctx, fmt.Sprintf("%s-listener", name), args, listOpts...)

	if err != nil {
		return nil, err
	}

	return listener, nil
}

func (l *PulumiInternalLoadBalancer) AttachLoadBalancer(ctx *pulumi.Context, name string, tgArn pulumi.StringOutput, opts ...pulumi.ResourceOption) error {

	return nil
}

type PulumiInternalLoadBalancer struct {
	pulumi.ResourceState

	LoadBalancer *lb.LoadBalancer
}

File: ecs-hosted/ts/application/networking/pulumiInternalLoadBalancer.ts
import * as pulumi from "@pulumi/pulumi";
import * as lb from "@pulumi/aws/lb";

import { LoadBalancerArgs } from "./types";

const namespace = "pulumi:internalLoadBalancer";

export class PulumiInternalLoadBalancer extends pulumi.ComponentResource {

    public readonly loadBalancer: lb.LoadBalancer;

    private readonly baseOptions: pulumi.ComponentResourceOptions;

    constructor(name: string, args: LoadBalancerArgs, opts?: pulumi.ComponentResourceOptions) {
        super(namespace, name, args, opts);

        this.baseOptions = pulumi.mergeOptions(opts, { parent: this, deleteBeforeReplace: true });

        const idleTimeout = args.idleTimeout ? args.idleTimeout : 120;

        this.loadBalancer = new lb.LoadBalancer(`${name}-lb`, {
            loadBalancerType: "network",
            internal: true,
            subnets: args.privateSubnetIds,
            idleTimeout: idleTimeout,
            ipAddressType: "ipv4",
        }, this.baseOptions);
    }

    createListener(name: string, tgArn: pulumi.Output<string>, certArn?: string, opts?: pulumi.ResourceOption[]): lb.Listener {
        const listenerArgs: lb.ListenerArgs = {
            loadBalancerArn: this.loadBalancer.arn,
            port: 80,
            protocol: "TCP",
            defaultActions: [{
                targetGroupArn: tgArn,
                type: "forward",
            }],
        };

        if (certArn) {
            listenerArgs.certificateArn = certArn;
            listenerArgs.sslPolicy = "ELBSecurityPolicy-TLS-1-2-2017-01";
            listenerArgs.protocol = "TLS";
            listenerArgs.port = 443;
        }

        const listOpts = pulumi.mergeOptions(this.baseOptions, { deleteBeforeReplace: true }, ...(opts || []));
        
        return new lb.Listener(`${name}-listener`, listenerArgs, listOpts);
    }
}